src/client/nuclearnet/web_socket_proxy_nuclearnet_client.ts

```diff
@@ -6,14 +6,21 @@ import { NUClearNetClient } from '../../shared/nuclearnet/nuclearnet_client'
 import { WebSocketClient } from './web_socket_client'
 import SocketIOSocket = SocketIOClient.Socket
 
+/**
+ * A client-side interface for interacting with NUClearNet. Allows a browser to connect transparently connect to
+ * NUClearNet, using web sockets at the transport layer. Supports automatic reconnection, which rebinds all NUClearNet
+ * event listeners.
+ */
 export class WebSocketProxyNUClearNetClient implements NUClearNetClient {
-  private nextMessageId: number
+  private nextRequestToken: number
+
+  // Store all listeners so that we can restore them after a socket disconnection/reconnection.
   private joinListeners: Set<NUClearEventListener>
   private leaveListeners: Set<NUClearEventListener>
-  private packetListeners: Map<string, Set<{ messageId: string, listener: NUClearPacketListener }>>
+  private packetListeners: Map<string, Set<{ requestToken: string, listener: NUClearPacketListener }>>
 
   public constructor(private socket: WebSocketClient) {
-    this.nextMessageId = 0
+    this.nextRequestToken = 0
```

Comment:
The problem with this is that the unsubscribe function (i.e. the `off()`) must be returned from `on()` synchronously. Getting an off method from the server will be async and can't be returned 😞 .